src/App.tsx

```diff
@@ -1,223 +1,173 @@
+    // eslint-disable-next-line react-hooks/exhaustive-deps
   }, []);
 
+  const onDeleteTodo = (todoId: number) => {
+    setDeletingTodoIds(prev => [...prev, todoId]);
+    deleteTodo(todoId)
+      .then(() => (
+        setTodos(allTodos => allTodos.filter(todo => todo.id !== todoId))
+      ))
+      .catch(() => {
+        setErrorMessage('Unable to delete a todo');
+      })
+      .finally(() => (
+        setDeletingTodoIds(prev => prev.filter(id => id !== todoId))));
+  };
+
+  const onFormSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    if (!title.trim()) {
+      setErrorMessage('Title can\'t be empty');
+
+      return;
+    }
+
+    if (user) {
+      setTempTodo({
+        id: 0,
+        title,
+        completed: false,
+        userId: user.id,
+      });
+      setIsAdding(true);
+
+      createTodo(title, user.id)
+        .then(newTodo => {
+          setTodos(prev => [...prev, {
+            id: newTodo.id,
+            userId: newTodo.userId,
+            title: newTodo.title,
+            completed: newTodo.completed,
+          }]);
+        })
+        .catch(() => setErrorMessage('Unable to add a todo'))
+        .finally(() => {
+          setIsAdding(false);
+          setTitle('');
+          setTempTodo(null);
+        });
+    }
+  };
+
+  const visibleTodos = useMemo(() => {
+    return todos.filter(todo => {
+      switch (completedFilter) {
+        case FilterType.ACTIVE:
+          return !todo.completed;
+
+        case FilterType.COMPLETED:
+          return todo.completed;
+
+        default:
+          return todos;
+      }
+    });
+  }, [completedFilter, todos]);
+
+  const uncompletedTodosLength = useMemo(() => {
+    const unfinishedTodos = todos.filter(todo => !todo.completed);
+
+    return unfinishedTodos.length;
+  }, [todos]);
+
+  const completedTodosLength = useMemo(() => {
+    const finishedTodos = todos.filter(todo => todo.completed);
+
+    return finishedTodos.length;
+  }, [todos]);
+
+  const clearCompletedButton = useCallback(() => {
+    todos.forEach(todo => {
+      if (todo.completed) {
+        onDeleteTodo(todo.id);
+      }
+    });
+  }, [todos]);
+
   return (
     <div className="todoapp">
       <h1 className="todoapp__title">todos</h1>
 
       <div className="todoapp__content">
-        <header className="todoapp__header">
-          <button
-            data-cy="ToggleAllButton"
-            type="button"
-            className="todoapp__toggle-all active"
-          />
-
-          <form>
-            <input
-              data-cy="NewTodoField"
-              type="text"
-              ref={newTodoField}
-              className="todoapp__new-todo"
-              placeholder="What needs to be done?"
+        <Header
+          newTodoField={newTodoField}
+          title={title}
+          onSetTitle={setTitle}
+          isAdding={isAdding}
+          onFormSubmit={onFormSubmit}
+        />
+        {todos.length !== 0 && (
```

Comment:
In this line should also check if there is a tempTodo in order to add the correct loader when adding the first todo
`{(todos.length !== 0 || tempTodo) && (`